pep-0682.rst

```diff
@@ -0,0 +1,183 @@
+PEP: 682
+Title: Format Specifier for Signed Zero
+Author: John Belmonte <john@neggie.net>
+Sponsor: Mark Dickinson <dickinsm@gmail.com>
+Discussions-To:
+Status: Draft
+Type: Standards Track
+Content-Type: text/x-rst
+Created: 29-Jan-2022
+Python-Version: 3.11
+Post-History:
+
+
+Abstract
+========
+
+Though ``float`` and ``Decimal`` types can represent `signed zero`_, in many 
+fields of mathematics, negative zero is surprising or unwanted -- especially
+in the context of displaying an (often rounded) numerical result.  This PEP
+proposes an extension to the `string format specification`_, allowing negative
+zero to be normalized to positive zero.
+
+.. _`signed zero`: https://en.wikipedia.org/wiki/Signed_zero
+.. _`string format specification`: https://docs.python.org/3/library/string.html#formatstrings
+
+
+Motivation
+==========
+
+Here is negative zero:
+
+.. code-block:: pycon
+
+    >>> x = -0.
+    >>> x
+    -0.0
+
+When formatting a number, negative zero can result from rounding.  Assuming
+the user's intention is truly to discard precision, the distinction between
+negative and positive zero of the rounded result might be considered an
+unwanted artifact:
+
+.. code-block:: pycon
+
+    >>> for x in (.002, -.001, .060):
+    ...     print(f'{x: .1f}')
+     0.0
+    -0.0
+     0.1
+
+There are various approaches to clearing the sign of a negative zero.  It
+can be achieved without a conditional by adding positive zero:
+
+.. code-block:: pycon
+
+    >>> x = -0.
+    >>> x + 0.
+    0.0
+
+To normalize negative zero when formatting, it is necessary to perform
+a redundant (and error-prone) pre-rounding of the input:
+
+.. code-block:: pycon
+
+    >>> for x in (.002, -.001, .060):
+    ...     print(f'{round(x, 1) + 0.: .1f}')
+     0.0
+     0.0
+     0.1
+
+What we would like instead is a first-class option to normalize negative
```

Comment:
FWIW, I wrote some notes a couple of weeks ago on this and had a couple of domain-oriented examples of my own. I'd be happy for those to be included if the author feels the might be useful or persuasive; if not, I'll save them for the discussion threads. (The first example would inevitably lead to a not-useful tangent of when using binary floating-point for representing monetary calculations is acceptable.)

```python
>>> previous_balance = 3.30
>>> payments = [1.10, 2.20]
>>> new_balance = previous_balance - sum(payments)
>>> print("Amount owed: {:.2f}".format(new_balance))
Amount owed: -0.00
```

and

```python
>>> celcius = -17.8
>>> fahrenheit = 32.0 + 1.8*celcius
>>> print(f"Today's temperature is {fahrenheit:.1f}°F.")
Today's temperature is -0.0°F.
```

Reply:
Thank you for the suggestion.  I think the existing rounding example gets at the heart of it by showing how precision intended to be discarded is affecting the output-- let's see if that's enough.

Yes, the `Decimal` type would certainly be warranted for an account that can only represent units of .01.

Specifically on the temperature example, one defense of displaying "-0" I came across was regarding temperature, where people really want to distinguish between below vs. precisely at 0 (especially in Celsius).